### DISCUSSION TITLE
MorphToMany with Multiple Morphs

### DISCUSSION THREAD

**Discussion by Zen0x7** [laravel/framework#40972] [Ideas]
Actually the framework doesn't support a relationships described on the title. 

Anyways, i wanna suggest some idea.

When you have an application and that application starts to growth, the database maintenance starts to be a problem, specially when you're a code optimizer and monolith creator. 

> What about a morphToMany relationships with multiple morphs?

Given the following scenario:

```
# Models

- User
- Group
- Vehicle
- House 

# Tables

- users
- groups
- vehicles
- houses
- relations

# Relations Table Columns

- parent_id
- parent_type
- child_id
- child_type
- created_at
- updated_at

# Queries

User::with(['groups', 'vehicles', 'houses'])
```

> What problem i'm trying to solve?

Well, it's very simple. In the other case you should have `group_user`, `user_vehicle`, `house_vehicle`  tables _and more_.

> How to solve them?

A single table of relations. With double morph columns.

Extends MorphsToMany or create a new relationship to supports double morphs in pivot table.

> What about pivot columns? If the models relationships shares a same contract then that would not be a problem. Sometimes you only need a simple link relationships between models.

**Comment by zachiler** [comment]
I'm trying to do exactly this, but it seems not possible without manually managing the morphed relationships. I got as far as SQL telling me that entity_assignable_type doesn't have a default value, but all other values are passed correctly when using:

```
public function locations(): MorphToMany
    {
        return $this->morphToMany(Location::class, 'assignable', 'assignables', 'assignable_id', 'entity_assignable_id')
            ->using(Assignable::class)
            ->withPivot('type', 'primary')
            ->withTimestamps();
    }
 ```

`$staff->locations()->attach($locationId);`

My only option right now is to split up the tables (like you've already mentioned) in order to ensure that I can use the attach() / detach() methods. While this isn't exactly a problem, it will create an additional tables for each assignment type I want to have.

Did you ever find any success?

**Reply by Zen0x7** [reply]
Hi @zachiler 

After 3 years. Saddly nope. I remember trying a few things, but they were too limited. Nothing like what the relationships currently implemented in Laravel offer. The idea behind this thread is a universal attachment of models. It sounds beautiful, but it's an architectural nightmare.

There are a lot of problems behind the idea ... i don't have time to explain everything ...

**Reply by zachiler** [reply]
Thanks for the reply. I agree there are a lot of things one must consider before attempting to implement something like this. I'm going to do it eventually though, I just don't have time to work it out at the moment.

**Comment by browner12** [comment]
you can currently fake this relationship by using the `withPivotValue()` method.

```php
/**
 * Retrieve the "tags" relationship.
 */
public function tags(): MorphToMany
{
    return $this->morphToMany(Tag::class, 'taggable', relatedPivotKey: 'tagger_id')
        ->withPivotValue('tagger_type', Tag::class);
}
```

I also have a branch currently in progress to bring this as a first-party feature to the framework, just gotta get off my butt and finish it.
